pallets/liquid-staking-two/src/lib.rs

```diff
@@ -247,45 +249,116 @@ pub mod pallet {
+        ) -> DispatchResultWithPostInfo {
+            let _sender = ensure_signed(origin)?;
+            MatchingPool::<T>::try_mutate(
+                Self::current_era(),
+                |matching_buffer| -> DispatchResult {
+                    let total_stake_amount = matching_buffer
+                        .total_stake_amount
+                        .checked_add(amount)
+                        .ok_or(ArithmeticError::Overflow)?;
+                    *matching_buffer = MatchingBuffer {
+                        total_stake_amount,
+                        ..matching_buffer.clone()
+                    };
+                    Ok(().into())
+                },
+            )?;
+
+            StakingOperationHistory::<T>::insert(
+                Self::current_era(),
+                StakingOperationType::RecordReward,
+                Operation {
+                    status: ResponseStatus::Successed,
+                    amount,
+                    block_number,
+                },
+            );
             Ok(().into())
         }
 
+        /// Invoked by offchain stake client. Declare reward received.
+        ///
+        /// `origin` should be approved account.(not implement yet).
+        ///
+        /// Q: Should update exchange_rate?
         #[pallet::weight(10_000)]
         #[transactional]
-        pub fn record_rebond_response(origin: OriginFor<T>) -> DispatchResultWithPostInfo {
+        pub fn record_slash(
+            origin: OriginFor<T>,
+            agent: T::AccountId,
+            #[pallet::compact] amount: Balance,
+            block_number: BlockNumber,
+        ) -> DispatchResultWithPostInfo {
+            let _sender = ensure_signed(origin)?;
+            MatchingPool::<T>::try_mutate(
+                Self::current_era(),
+                |matching_buffer| -> DispatchResult {
+                    let total_stake_amount = matching_buffer
+                        .total_stake_amount
+                        .checked_sub(amount)
+                        .ok_or(ArithmeticError::Underflow)?;
+                    *matching_buffer = MatchingBuffer {
+                        total_stake_amount,
+                        ..matching_buffer.clone()
+                    };
+                    Ok(().into())
+                },
+            )?;
+
+            StakingOperationHistory::<T>::insert(
+                Self::current_era(),
+                StakingOperationType::RecordSlash,
+                Operation {
+                    status: ResponseStatus::Successed,
+                    amount,
+                    block_number,
+                },
+            );
             Ok(().into())
         }
 
+        /// Invoked by offchain stake client.
+        ///
+        /// Record bond operation and corresponding data.
         #[pallet::weight(10_000)]
         #[transactional]
-        pub fn record_unbond_response(origin: OriginFor<T>) -> DispatchResultWithPostInfo {
+        pub fn record_bond_response(
```

Comment:
for this method, I think we need first to check if the `StakingOperationType` exists and its status is not successful,

Reply:
Could you please explain? I'm kind of confused about it. Should there be only one bonding operation in an era?

Reply:
yes, each era can only have one bonding or unbonding operation